File: memind-clients/typescript/tests/retry.test.ts
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { describe, expect, it, vi } from 'vitest'
import { MemindClient } from '../src/client.js'
import { computeRetryDelay, isRetryableStatus, parseRetryAfter } from '../src/core/retry.js'
import { Message, RawContent } from '../src/types/message.js'

describe('isRetryableStatus', () => {
  it('returns true for retryable codes', () => {
    for (const code of [408, 429, 500, 502, 503, 504]) {
      expect(isRetryableStatus(code)).toBe(true)
    }
  })

  it('returns false for non-retryable codes', () => {
    for (const code of [200, 400, 401, 403, 404]) {
      expect(isRetryableStatus(code)).toBe(false)
    }
  })
})

describe('computeRetryDelay', () => {
  it('returns initial delay for attempt 0', () => {
    const delay = computeRetryDelay(0)
    expect(delay).toBeGreaterThanOrEqual(375)
    expect(delay).toBeLessThanOrEqual(625)
  })

  it('caps at max delay', () => {
    const delay = computeRetryDelay(10)
    expect(delay).toBeLessThanOrEqual(2500)
  })

  it('uses retry-after without capping server-requested delay', () => {
    expect(computeRetryDelay(0, 10_000)).toBe(10_000)
  })
})

describe('parseRetryAfter', () => {
  it('parses integer seconds', () => {
    expect(parseRetryAfter('30')).toBe(30000)
  })

  it('returns undefined for invalid values', () => {
    expect(parseRetryAfter(null)).toBeUndefined()
    expect(parseRetryAfter('')).toBeUndefined()
  })

  it('parses HTTP-date format', () => {
    const future = new Date(Date.now() + 5000).toUTCString()
    const result = parseRetryAfter(future)
    expect(result).toBeGreaterThan(3000)
    expect(result).toBeLessThanOrEqual(6000)
  })
})

describe('retry integration', () => {
  it('retrieve retries on 503', async () => {
    let calls = 0
    const mockFetch = vi.fn().mockImplementation(() => {
      calls++
      if (calls <= 2) {
        return Promise.resolve({
          status: 503,
          headers: new Headers({ 'content-type': 'application/json' }),
          json: async () => ({ error: { code: 'unavailable', message: 'down' } }),
        })
      }
      return Promise.resolve({
        status: 200,
        headers: new Headers({ 'content-type': 'application/json' }),
        json: async () => ({
          data: { status: 'OK', items: [], insights: [], rawData: [], evidences: [] },
        }),
      })
    })

    const client = new MemindClient({ baseUrl: 'http://x', fetch: mockFetch, maxRetries: 2 })
    const result = await client.memory.retrieve({
      userId: 'u1',
      agentId: 'a1',
      query: 'test',
      strategy: 'SIMPLE',
    })
    expect(result.items).toEqual([])
    expect(calls).toBe(3)
  })

  it('mutating ops do NOT retry by default', async () => {
    let calls = 0
    const mockFetch = vi.fn().mockImplementation(() => {
      calls++
      return Promise.resolve({
        status: 503,
        headers: new Headers({ 'content-type': 'application/json' }),
        json: async () => ({ error: { code: 'unavailable', message: 'down' } }),
      })
    })

    const client = new MemindClient({ baseUrl: 'http://x', fetch: mockFetch, maxRetries: 2 })
    await expect(
      client.memory.addMessage({
        userId: 'u1',
        agentId: 'a1',
        message: Message.user('hi'),
      }),
    ).rejects.toThrow()
    expect(calls).toBe(1)
  })

  it('mutating ops retry when per-request maxRetries is set', async () => {
    let calls = 0
    const mockFetch = vi.fn().mockImplementation(() => {
      calls++
      if (calls <= 1) {
        return Promise.resolve({
          status: 503,
          headers: new Headers({ 'content-type': 'application/json' }),
          json: async () => ({ error: { code: 'unavailable', message: 'down' } }),
        })
      }
      return Promise.resolve({
        status: 200,
        headers: new Headers({ 'content-type': 'application/json' }),
        json: async () => ({ data: { triggered: false } }),
      })
    })

    const client = new MemindClient({ baseUrl: 'http://x', fetch: mockFetch })
    await client.memory.addMessage(
      { userId: 'u1', agentId: 'a1', message: Message.user('hi') },
      { maxRetries: 1 },
    )
    expect(calls).toBe(2)
  })

  it('per-request maxRetries overrides client defaults for retrieve', async () => {
    let calls = 0
    const mockFetch = vi.fn().mockImplementation(() => {
      calls++
      return Promise.resolve({
        status: 503,
        headers: new Headers({ 'content-type': 'application/json' }),
        json: async () => ({ error: { code: 'unavailable', message: 'down' } }),
      })
    })

    const client = new MemindClient({ baseUrl: 'http://x', fetch: mockFetch, maxRetries: 2 })
    await expect(
      client.memory.retrieve(
        { userId: 'u1', agentId: 'a1', query: 'test', strategy: 'SIMPLE' },
        { maxRetries: 0 },
      ),
    ).rejects.toThrow()
    expect(calls).toBe(1)
  })

  it('per-request timeoutMs overrides the client default', async () => {
    const client = new MemindClient({
      baseUrl: 'http://x',
      timeoutMs: 10_000,
      fetch: vi.fn().mockImplementation(
        (_url, opts: RequestInit) =>
          new Promise((_resolve, reject) => {
            opts.signal?.addEventListener('abort', () => {
              const err = new Error('aborted')
              err.name = 'AbortError'
              reject(err)
            })
          }),
      ),
    })

    await expect(client.health({ timeoutMs: 1 })).rejects.toThrow(/timed out/)
  })

  it('invalid raw content is rejected before network calls', async () => {
    const mockFetch = vi.fn()
    const client = new MemindClient({ baseUrl: 'http://x', fetch: mockFetch })

    expect(() => RawContent.conversation([])).toThrow()
    expect(() =>
      RawContent.map('document', {
        title: 'ok',
        bad: undefined as never,
      }),
    ).toThrow(/JSON-serializable/)

    const rawContent = {
      type: 'document',
      title: 'ok',
      bad: undefined,
    } as never
    await expect(
      client.memory.extract({
        userId: 'u1',
        agentId: 'a1',
        rawContent,
      }),
    ).rejects.toThrow(/non-serializable|unsupported|undefined|serializable/)
    expect(mockFetch).not.toHaveBeenCalled()
  })
})

File: memind-clients/rust/tests/retry_test.rs
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use memind::{
    AddMessageRequest, CommitMemoryRequest, ExtractMemoryRequest, MemindClient, RawContent,
    RequestOptions, RetrieveMemoryRequest,
};
use serde_json::json;
use wiremock::matchers::{method, path};
use wiremock::{Mock, MockServer, ResponseTemplate};

#[tokio::test]
async fn retrieve_retries_on_503_by_default() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path("/open/v1/memory/retrieve"))
        .respond_with(
            ResponseTemplate::new(503)
                .append_header("Retry-After", "0")
                .set_body_json(json!({
                    "error": {"code": "unavailable", "message": "try again"}
                })),
        )
        .expect(3)
        .mount(&server)
        .await;

    let client = MemindClient::builder()
        .base_url(server.uri())
        .build()
        .unwrap();
    let err = client
        .memory()
        .retrieve(RetrieveMemoryRequest::new(
            "u1",
            "a1",
            "what",
            memind::Strategy::Simple,
        ))
        .await
        .unwrap_err();
    assert!(matches!(err, memind::MemindError::Api(_)));
}

#[tokio::test]
async fn health_retries_on_503_by_default() {
    let server = MockServer::start().await;
    Mock::given(method("GET"))
        .and(path("/open/v1/health"))
        .respond_with(
            ResponseTemplate::new(503)
                .append_header("Retry-After", "0")
                .set_body_json(json!({
                    "error": {"code": "unavailable", "message": "try again"}
                })),
        )
        .expect(3)
        .mount(&server)
        .await;

    let client = MemindClient::builder()
        .base_url(server.uri())
        .build()
        .unwrap();
    let err = client.health().await.unwrap_err();
    assert!(matches!(err, memind::MemindError::Api(_)));
}

#[tokio::test]
async fn add_message_does_not_retry_by_default() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path("/open/v1/memory/sync/add-message"))
        .respond_with(ResponseTemplate::new(503).set_body_json(json!({
            "error": {"code": "unavailable", "message": "try again"}
        })))
        .expect(1)
        .mount(&server)
        .await;

    let client = MemindClient::builder()
        .base_url(server.uri())
        .build()
        .unwrap();
    let err = client
        .memory()
        .add_message(AddMessageRequest::new(
            "u1",
            "a1",
            memind::Message::user("hello"),
        ))
        .await
        .unwrap_err();
    assert!(matches!(err, memind::MemindError::Api(_)));
}

#[tokio::test]
async fn extract_and_commit_do_not_retry_by_default() {
    let server = MockServer::start().await;
    for endpoint in [
        "/open/v1/memory/sync/extract",
        "/open/v1/memory/sync/commit",
    ] {
        Mock::given(method("POST"))
            .and(path(endpoint))
            .respond_with(ResponseTemplate::new(503).set_body_json(json!({
                "error": {"code": "unavailable", "message": "try again"}
            })))
            .expect(1)
            .mount(&server)
            .await;
    }

    let client = MemindClient::builder()
        .base_url(server.uri())
        .build()
        .unwrap();
    let raw = RawContent::conversation(vec![memind::Message::user("remember")]).unwrap();
    let extract_err = client
        .memory()
        .extract(ExtractMemoryRequest::new("u1", "a1", raw))
        .await
        .unwrap_err();
    assert!(matches!(extract_err, memind::MemindError::Api(_)));

    let commit_err = client
        .memory()
        .commit(CommitMemoryRequest::new("u1", "a1"))
        .await
        .unwrap_err();
    assert!(matches!(commit_err, memind::MemindError::Api(_)));
}

#[tokio::test]
async fn mutating_operation_retries_when_per_request_max_retries_is_set() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path("/open/v1/memory/sync/commit"))
        .respond_with(
            ResponseTemplate::new(503)
                .append_header("Retry-After", "0")
                .set_body_json(json!({
                    "error": {"code": "unavailable", "message": "try again"}
                })),
        )
        .expect(2)
        .mount(&server)
        .await;

    let client = MemindClient::builder()
        .base_url(server.uri())
        .build()
        .unwrap();
    let err = client
        .memory()
        .commit_with_options(
            CommitMemoryRequest::new("u1", "a1"),
            RequestOptions::new().max_retries(1),
        )
        .await
        .unwrap_err();
    assert!(matches!(err, memind::MemindError::Api(_)));
}

#[tokio::test]
async fn per_request_max_retries_zero_disables_retrieve_retry() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path("/open/v1/memory/retrieve"))
        .respond_with(ResponseTemplate::new(503).set_body_json(json!({
            "error": {"code": "unavailable", "message": "try again"}
        })))
        .expect(1)
        .mount(&server)
        .await;

    let client = MemindClient::builder()
        .base_url(server.uri())
        .build()
        .unwrap();
    let err = client
        .memory()
        .retrieve_with_options(
            RetrieveMemoryRequest::new("u1", "a1", "what", memind::Strategy::Simple),
            RequestOptions::new().max_retries(0),
        )
        .await
        .unwrap_err();
    assert!(matches!(err, memind::MemindError::Api(_)));
}
